src/components/z-popover/index.tsx

```diff
@@ -0,0 +1,151 @@
+  @Prop() showArrow?: boolean = false;
+
+  @State() isVisible: boolean = false;
+
+  private popoverElem: HTMLElement;
+  private defaultPosition: PopoverPosition = this.position;
+
+  @Listen("closePopover")
+  closePopover() {
+    this.isVisible = false;
+  }
+
+  @Listen("keyup", { target: "window" })
+  closePopoverWithKeyboard(e: any) {
+    if (e.key === KeyboardKeys.ESC) {
+      this.closePopover();
+    }
+  }
+
+  handleClick() {
+    this.isVisible = !this.isVisible;
+    this.checkSpaceAvailable();
+  }
+
+  @Listen("click", { target: "body", capture: true })
+  handleOutsideClick() {
+    const tree = getElementTree(document.activeElement);
+    const parent = tree.find(
+      (elem: any) => elem.nodeName.toLowerCase() === "z-popover"
+    );
+
+    if (!parent) {
+      this.isVisible = false;
+    }
+  }
+
+  checkSpaceAvailable() {
+    if (!this.isVisible) {
+      this.position = this.defaultPosition;
+      return;
+    }
+
+    const width = document.body.clientWidth;
+    const height = window.innerHeight;
+
+    const l = this.popoverElem.getBoundingClientRect().left;
+    const r = this.popoverElem.getBoundingClientRect().right;
+    const t = this.popoverElem.getBoundingClientRect().top;
+    const b = this.popoverElem.getBoundingClientRect().bottom;
+
+    let firstSide = this.position.split("-")[0];
+    let secondSide = this.position.split("-")[1];
+
+    // If top is outside viewport
+    if (t < 0) {
+      if (this.position.startsWith("above")) {
+        firstSide = "below";
+      } else {
+        secondSide = "down";
+      }
+    }
+
+    // If bottom is outside viewport
+    if (b > height) {
+      if (this.position.startsWith("below")) {
+        firstSide = "above";
+      } else {
+        secondSide = "up";
+      }
+    }
+
+    // If right is outside viewport
+    if (r > width) {
+      if (
+        this.position.startsWith("above") ||
+        this.position.startsWith("below")
+      ) {
+        secondSide = "left";
+      } else {
+        firstSide = "before";
+      }
+    }
+
+    // If left is outside viewport
+    if (l < 0) {
+      if (
+        this.position.startsWith("above") ||
+        this.position.startsWith("below")
+      ) {
+        secondSide = "right";
+      } else {
+        firstSide = "after";
+      }
+    }
+
+    this.position = PopoverPosition[`${firstSide}-${secondSide}`];
+  }
+
+  render() {
+    return (
+      <Host>
+        <div onClick={() => this.handleClick()}>
+          <slot name="trigger"></slot>
+        </div>
+        <div
+          ref={(e) => (this.popoverElem = e)}
+          class={classNames(
+            "popover-content-container",
+            this.position,
+            `border-radius-${this.borderRadius}`,
+            this.boxShadow,
+            { "show-arrow": this.showArrow },
+            { visible: this.isVisible }
+          )}
+          style={{
+            backgroundColor: `var(--${this.backgroundColor})`,
```

Comment:
Here we could use a fallback (to 'white') in case the value passed in the prop in not a defined variable 
[documentation](https://developer.mozilla.org/en-US/docs/Web/CSS/var()#custom_properties_with_fallbacks_for_use_when_the_property_has_not_been_set)

Reply:
Di questo ne avevamo parlato in chat. Al momento vengono gestite solo come prop e se la prop token non è valida non viene gestita la proprietà. Comunque sarà un argomento di discussione di mercoledì